### PR TITLE
Avoid clone for sparse tensors during accumulation of grads.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33427 Avoid clone for sparse tensors during accumulation of grads.**
* #33214 Unify gradient accumulation between distributed autograd and local autograd
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -3240,6 +3240,78 @@ def backward(ctx, grad):
...
+        p_g = MyFunc.static_grad_ptr
+        p_a = a.grad._values().data_ptr()
+        p_b = b.grad._values().data_ptr()
+        # check a,b uses different grad buffer
+        self.assertFalse(p_a == p_b)
+        # check one of them is using the computed buffer
+        self.assertTrue(p_a == p_g or p_b == p_g)
+
+        # Run backwards multiple times to ensure accumulation works.
+        for i in range(100):
```
Comment (albanD) [inline test/test_autograd.py]: Maybe 10 is enough here? We want to keep the runtime of tests low.

File: test/test_autograd.py [hunk 2]
```diff
@@ -3240,6 +3240,78 @@ def backward(ctx, grad):
...
+            loss.backward(retain_graph=True)
+
+        # non-contiguous indices and value should not trigger copy either.
```
Comment (albanD) [inline test/test_autograd.py]: Maybe they should? The current behavior would return contiguous indices and values all the time no?

File: torch/testing/_internal/distributed/rpc/dist_autograd_test.py [hunk 3]
```diff
@@ -1869,6 +1869,82 @@ def backward(ctx, grad):
...
+            p_g = MyFunc.static_grad_ptr
+            p_a = grads[a]._values().data_ptr()
+            p_b = grads[b]._values().data_ptr()
+            # check a,b uses different grad buffer
+            self.assertFalse(p_a == p_b)
+            # check one of them is using the computed buffer
+            self.assertTrue(p_a == p_g or p_b == p_g)
+
+            # Run backwards multiple times.
+            for i in range(100):
```
Comment (albanD) [inline torch/testing/_internal/distributed/rpc/dist_autograd_test.py]: Same comment as above for the number of iterations.

**Review by albanD — COMMENTED** [review]
As discussed offline, if we want to double check if this is BC breaking or not.
So that we either make it BC or we mark the PR as BC-breaking.

File: test/test_autograd.py (hunk 2 above)
Comment (pritamdamania87) [reply test/test_autograd.py]: I'm not sure about this, is this a contract we make with users currently? If not it seems like we should optimize this codepath as well. Are there any ops on SparseTensors that fail if indices and values are non-contiguous? Based on what I saw, we usually handle non-contiguous indices and values in SparseTensors appropriately.

Comment (albanD) [reply test/test_autograd.py]: I'm afraid there is no contract :/ So anything that changes the current behavior is a BC-breaking change.
A user could be giving this directly into a custom cpp function that assumes that indices and values are contiguous for example.

File: torch/testing/_internal/distributed/rpc/dist_autograd_test.py [hunk 4]
```diff
@@ -1898,3 +1898,80 @@ def backward(ctx, grad):
...
+            # check a,b uses different grad buffer
+            self.assertFalse(p_a == p_b)
+            # check one of them is using the computed buffer
+            self.assertTrue(p_a == p_g or p_b == p_g)
+
+            # Run backwards multiple times.
+            for i in range(10):
+                dist_autograd.backward([loss], retain_graph=True)
+
+        # non-contiguous indices and value, we should trigger a copy.
```
Comment (gqchen) [inline torch/testing/_internal/distributed/rpc/dist_autograd_test.py]: do we want to test contiguous indices and values?

**Review by gqchen — COMMENTED** [review]
one comment on the test. Otherwise it looks good to me. 

@albanD, can you elaborate on why we think this might be BC breaking? The sparse new grad by definition is not referenced anywhere else.

**Comment by albanD** [comment]
The BC-breaking change is about the state of the `_values` and `_indices` Tensors inside the SparseTensor. In particular, since the old code always use `.clone()`, these Tensors are always "contiguous". This is particularly important because if someone tries to modify these inplace, the behavior can be unexpected.

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 5]
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
         // holding references to the Tensor and that is fine since these are not
         // exposed to the user.
         update_grad(new_grad_copy.detach());
+      } else if (
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
+          new_grad_copy._values().is_contiguous() &&
```
Comment (gqchen) [inline torch/csrc/autograd/functions/accumulate_grad.h]: @albanD Maybe I missed something obvious. We do shallow copy only when indices and values are contiguous?

**Review by gqchen — COMMENTED** [review]
Maybe I missed something obvious. In this version of the code, we do shallow copy only when _values and _indices are contiguous?

File: torch/csrc/autograd/functions/accumulate_grad.h (hunk 5 above)
Comment (albanD) [reply torch/csrc/autograd/functions/accumulate_grad.h]: Yes. If they are already contiguous and not used anywhere else, it is fine to reuse the same memory.
Otherwise, we use clone to get brand new memory.

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 6]
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
...
+          new_grad_use_count <= num_expected_refs) {
```
Comment (albanD) [inline torch/csrc/autograd/functions/accumulate_grad.h]: Actually following the comment above. We want to check the use_count for the indices and values as well here no?

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 7]
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
         // holding references to the Tensor and that is fine since these are not
         // exposed to the user.
         update_grad(new_grad_copy.detach());
+      } else if (
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
```
Comment (pritamdamania87) [inline torch/csrc/autograd/functions/accumulate_grad.h]: We need to add a check to ensure sparse tensors are always created with continguous indicies and values: https://github.com/pytorch/pytorch/issues/12633. I'd like to keep the checks we have here and fix sparse tensor constructor and the logic here in a followup PR.

[the remark this answers is not in the document]
Comment (albanD) [reply torch/csrc/autograd/functions/accumulate_grad.h]: This should be a different PR I think.
So that it is easier for us to measure the impact of doing this and add the corresponding tests.

[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/csrc/autograd/functions/accumulate_grad.h]: I see, is there an existing gh issue for this or should I create one?

Comment (pritamdamania87) [reply torch/csrc/autograd/functions/accumulate_grad.h]: @ezyang Was looking into some of the Sparse Tensor math and it seems like some operations do handle non-contiguous indices/values appropriately: https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/native/sparse/SparseTensorMath.cpp.html#_ZN2at6native18add_out_sparse_cpuERNS_6TensorERKS1_S4_N3c106ScalarE. I'm wondering if its actually fine to build SparseTensors with non-contiguous indices/values?

Comment (ezyang) [reply torch/csrc/autograd/functions/accumulate_grad.h]: Even if it is fine, I assure you that it is (1) not tested and (2) doesn't work on a bunch of kernels. And unlike noncontiguous dense tensors, where there is a clear value add for non-contiguity (views), there is not really any clear value add to handling these cases correctly. For example, in CUDA:

```
  if (is_same_tensor(r, dense_buffer)) {
    TORCH_CHECK(r_.is_contiguous(), "add: CUDA dense-sparse addition with a non-contiguous output tensor does not work; shout if you need it (see https://github.com/pytorch/pytorch/issues/1521 )");
```

Comment (ezyang) [reply torch/csrc/autograd/functions/accumulate_grad.h]: It might be possible to "best effort" non-contiguous values/indices by sprinkling contiguous everywhere in sparse but it's not clear to me that this is actually worth it. cc'ing @kurtamohler @nikitaved @hameerabbasi  for any other thoughts.

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 8]
```diff
@@ -50,6 +50,26 @@ struct TORCH_API AccumulateGrad : public Node {
...
+          new_grad._values().use_count() <= num_expected_refs &&
+          new_grad.use_count() <= num_expected_refs) {
+        // Can't detach sparse tensor (since metadata changes are not allowed
+        // after detach), so just create a new one for the grad which is a
+        // shallow copy. We need a shallow copy so that modifying the original
+        // grad tensor doesn't modify the grad we accumulate.
+        // We only skip clone if indices and values themselves are contiguous
+        // for backward compatiblity reasons. Since without this optimization,
+        // earlier we would clone the entire SparseTensor which cloned indices
+        // and values.
```
Comment (ezyang) [inline torch/csrc/autograd/functions/accumulate_grad.h]: ISTR, weren't you going to file a GitHub issue for the `detach()` issue and link it here?

File: torch/testing/_internal/distributed/rpc/dist_autograd_test.py [hunk 9]
```diff
@@ -1932,3 +1932,80 @@ def backward(ctx, grad):
             self.assertFalse(p_a == p_b)
             # check one of them is using the computed buffer
             self.assertTrue(p_a == p_g or p_b == p_g)
+
+    @dist_init
+    def test_no_grad_copy_sparse(self):
```
Comment (ezyang) [inline torch/testing/_internal/distributed/rpc/dist_autograd_test.py]: It's not so nice that the test has to be copy pasted like this. It would be good if at least the two versions of the test are cross-linked from each other.

**Review by ezyang — APPROVED** [review]

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 10]
```diff
@@ -50,6 +50,26 @@ struct TORCH_API AccumulateGrad : public Node {
         // holding references to the Tensor and that is fine since these are not
         // exposed to the user.
         update_grad(new_grad.detach());
+      } else if (
+          !GradMode::is_enabled() && new_grad.is_sparse() &&
+          new_grad._indices().is_contiguous() &&
+          new_grad._values().is_contiguous() &&
+          new_grad._indices().use_count() <= num_expected_refs &&
```
Comment (albanD) [inline torch/csrc/autograd/functions/accumulate_grad.h]: Is that correct? Shouldn't `num_expected_refs` be 1 here?

File: torch/testing/_internal/distributed/rpc/dist_autograd_test.py (hunk 9 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/rpc/dist_autograd_test.py]: I originally thought about doing this, but the distributed autograd tests are different enough (with the context, a different API and the need to use get_gradients for the grads map) that I wasn't sure how much we would be able to dedup effectively.

Additionally since we removed the optimization of the same grad tensor being used in different variables, the behavior between distributed autograd and local autograd for the same test is slightly different.

File: torch/csrc/autograd/functions/accumulate_grad.h (hunk 7 above)
Comment (hameerabbasi) [reply torch/csrc/autograd/functions/accumulate_grad.h]: If we're using external libraries in PyTorch, like MKL, those basically demand that we have contiguous values/indices. All literature I've read also implies this indirectly.

Although, of course, we could use non-contiguous versions, that'd be a huge performance hit over the already significant algorithmic hit from using sparse arrays of the same `nnz` as dense arrays. I might be wrong here, but it'd also be bad for GPU sparse arrays.

So, yes, I agree with the sentiment that ensuring contiguousness is a best practice for sparse arrays.

File: torch/csrc/autograd/functions/accumulate_grad.h (hunk 10 above)
Comment (albanD) [reply torch/csrc/autograd/functions/accumulate_grad.h]: These will have extra refs to the sparse Tensors. But the indices and values Tensors should only be referenced by the sparse Tensor.

Comment (pritamdamania87) [reply torch/csrc/autograd/functions/accumulate_grad.h]: Yes, you are right. I'll fix this in a follow up PR since this one is already landing.

Comment (gchanan) [reply torch/csrc/autograd/functions/accumulate_grad.h]: how do you know indices and values only have 1 refcount?  What if you have a pre-existing reference to the indices tensor, from before you created the sparse tensor?

Comment (gchanan) [reply torch/csrc/autograd/functions/accumulate_grad.h]: ah ok, *checking* 1 should be okay.

File: test/test_autograd.py [hunk 11]
```diff
@@ -3235,6 +3235,79 @@ def backward(ctx, grad):
...
+        p_a = a.grad._values().data_ptr()
+        p_b = b.grad._values().data_ptr()
+        # check a,b uses different grad buffer
+        self.assertFalse(p_a == p_b)
+        # check one of them is using the computed buffer
+        self.assertTrue(p_a == p_g or p_b == p_g)
+
+        # Run backwards multiple times to ensure accumulation works.
+        for i in range(10):
+            loss.backward(retain_graph=True)
```
Comment (gchanan) [inline test/test_autograd.py]: shouldn't we check the accumulation?  This should be easy because the grad_input is the same as the grad_output, right?
